### PR TITLE
Feature/highpass

### LINKED ISSUE #178: Develop a high pass filter algorithm
### Current Features

none

### Desired Additional Features

mirror of #173 but for high pass (and, the APIs of these should look very similarly, but this should happen by default based on our feedback)

In the middleware folder, we should create a simple high pass filter implementation. This should expose an initialization API for the filter (probably will include an alpha and scale factor inputs), and then it should have an API for filtering a sample of data

### Screenshots (as needed)

_No response_

### PR DESCRIPTION
**Pull request by rubennoro**
-Implemented High Pass Algorithm with scale and alpha factors
-Code doesn't build because of issues with PCA driver code, don't think those are relevant

_Any other notes go here_

## Test Cases

- Case A
- Edge case
- ...

## To Do

_Any remaining things that need to get done_

- [ ] item 1
- [ ] ...

## Checklist

It can be helpful to check the `Checks` and `Files changed` tabs.
Please reach out to your Project Lead if anything is unclear.
Please request reviewers and ping on slack only after you've gone through this whole checklist.

- [ ] All commits are tagged with the ticket number
- [ ] No merge conflicts
- [ ] All checks passing
- [ ] Remove any non-applicable sections of this template
- [ ] Assign the PR to yourself
- [ ] Request reviewers & ping on Slack
- [ ] PR is linked to the ticket (fill in the closes line below)

Closes #178 (issue #178 )

### CODE REVIEW

**Review by dyldonahue — CHANGES_REQUESTED** [review]
I think the algorithm is good ( i dont have correct ones memorized by heart, but this seems roughly correct so i trust you sourced it well. 

only issueS

1. There isnt any room to add a new data point. The way this would likely to be used is that u create the filter, and can over time, add a new data point (ie, one more buffer). If done correctly, you shouldnt need to store the value of SAMPLES worth of data. Just the current data point, and the summation of the output up until this (prev_sum)

 
2. Right now, could only b used in one place/for one thing. Else, if you try and have two high pass filters, the data will combine and the static variables will belnd together.

the best solution is to create a high pass filter struct object, and instead of using statci variables, you store those as members of the struct, that way we can create multiple instances of the struct.

Id create a "high pass filter init" function that creates and returns a high pass filter object with a given alpha and scale, and the current "high_pass" API u have would just take in the new data point and the hgih pass object which should give access to the static data needed

File: general/include/pca9539.h [hunk 1]
```diff
@@ -5,8 +5,7 @@
 #include <stdint.h>
 
 /*
-PCA 9539 16 bit GPIO expander.  Datasheet:
-https://www.ti.com/lit/ds/symlink/pca9539.pdf?ts=1716785085909
+PCA 9539 16 bit GPIO expander.  Datasheet: https://www.ti.com/lit/ds/symlink/pca9539.pdf?ts=1716785085909
```
Comment (Sabramz) [inline general/include/pca9539.h:8]: PCA changes shouldn't be in this PR

File: middleware/include/high_pass.h [hunk 2]
```diff
@@ -0,0 +1,8 @@
+#ifndef HIGH_PASS_H
+#define HIGH_PASS_H
+
+#include <stdio.h>
+
+int high_pass(int alpha, int scale);
```
Comment (Sabramz) [inline middleware/include/high_pass.h:6]: There are no functions for the user to pass a value through the filter.

**Review by Sabramz — CHANGES_REQUESTED** [review]

File: middleware/src/high_pass.c [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+//Samples can be modified
+#define SAMPLES 12
+
+//Scale Var allows adjustment of output strength
+//Alpha Var controls weight of prev input & sum difference
+float high_pass(int alpha, int scale)
+{
+	static float buffer[SAMPLES] = { 0 }; //initialize changing buffer
+	static int index = 0;
+	static float prev_sum = 0.0;
```
Comment (Sabramz) [inline middleware/src/high_pass.c:31]: This implementation does not allow for using multiple filters at once.

File: middleware/src/high_pass.c [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+*/
+#include "high_pass.h"
+
+//Samples can be modified
+#define SAMPLES 12
+
+//Scale Var allows adjustment of output strength
+//Alpha Var controls weight of prev input & sum difference
+float high_pass(int alpha, int scale)
```
Comment (Sabramz) [inline middleware/src/high_pass.c:27]: Should these values be ints? Floats might also be useful here.

File: middleware/include/high_pass.h (hunk 2 above)
Comment (Sabramz) [inline middleware/include/high_pass.h:6]: Documentation for the function should be in the header file and use doxygen, javadoc style comments.

File: middleware/src/high_pass.c [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+1. first possibility
+output = alpha * last_output + (1-alpha) * input
+
+output = alpha * last_output + alpha * (avg - prev_avg)
+
+*/
+#include "high_pass.h"
+
+//Samples can be modified
+#define SAMPLES 12
```
Comment (Sabramz) [inline middleware/src/high_pass.c:23]: Make this modifiable by the user. They should not have to edit code in embedded base. This also should be available to be edited on a per high-pass filter level.

**Review by Sabramz — CHANGES_REQUESTED** [review]

File: middleware/src/high_pass.c [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+
+output = alpha * last_output + alpha * (avg - prev_avg)
+
+*/
+
+#include "high_pass.h"
+
+void high_pass_init(float alpha, float scale, high_pass_st filter)
+{
+	filter->alpha = alpha;
```
Comment (Sabramz) [inline middleware/src/high_pass.c:25]: This is incorrect syntax and will not compile

Comment (Sabramz) [inline middleware/src/high_pass.c:25]: This is incorrect syntax and will not compile

**Review by Sabramz — CHANGES_REQUESTED** [review]

File: middleware/include/high_pass.h [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+
+} high_pass_t;
+
+/**
+ * @brief 
+ * 
+ * @param alpha 
+ * @param scale 
+ * @param filter 
+ */
```
Comment (Sabramz) [inline middleware/include/high_pass.h:21]: Need to fill in doc comments

**Review by Sabramz — CHANGES_REQUESTED** [review]

**Review by Sabramz — APPROVED** [review]

**Comment by Sabramz** [comment]
Looks good. Merge after thorough testing.

**Comment by Sabramz** [comment]
Outline the test cases tested in the header of this PR.
